src/aws_client.erl

```diff
@@ -1,34 +1,135 @@
+         get_creds/1]).
+
+-record(state, {
+         }).
 %%====================================================================
 %% API
 %%====================================================================
 
-make_client(AccessKeyID, SecretAccessKey, Region)
-  when is_binary(AccessKeyID), is_binary(SecretAccessKey), is_binary(Region) ->
-    #{access_key_id => AccessKeyID,
-      secret_access_key => SecretAccessKey,
-      region => Region,
-      endpoint => <<"amazonaws.com">>,
-      service => undefined}.
+start_link() ->
+  gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+stop() ->
+  gen_server:stop(?MODULE).
+
+make_client() ->
+  make_client(metadata, []).
+make_client(metadata, Opts) ->
+  gen_server:call(?MODULE, {add_creds, metadata, Opts}).
+% Backwards compatibility
+make_client(AccessKey, Secret, Region) when is_binary(Region) ->
+  make_client(static, {AccessKey, Secret}, [{region, Region}]);
+make_client(static, {AccessKey, Secret}, Opts) when is_list(Opts) ->
+  gen_server:call(?MODULE, {add_creds, {static, AccessKey, Secret}, Opts}).
+
+get_creds(Ref) ->
+  case ets:lookup(creds, Ref) of
+    [{_, Creds}] -> Creds;
+    [] -> undefined
+  end.
 
 %%====================================================================
-%% Unit tests
+%% Behaviour
 %%====================================================================
+init(_Args) ->
+  ets:new(creds, [set, named_table]),
+  ets:new(creds_dedup, [set, named_table]),
+  {ok, #state{}}.
+
+terminate(_Reason, _State) ->
+  ok.
+
+handle_call({add_creds, metadata, Opts}, _From, State) ->
+  case ets:lookup(creds_dedup, metadata) of
+    [{_, Ref}] -> {reply, {ok, Ref}, State};
+    [] ->
+      {ok, Role} = get_role(),
+      {ok, AccessKey, SecretKey, Expiry} = get_metadata_creds(Role),
+      {Region, Endpoint} = parse_opts(Opts),
+      Ref = make_ref(),
+      true = ets:insert(creds_dedup, {metadata, Ref}),
+      true = ets:insert(creds, {Ref, #{access_key => AccessKey,
+                                       secret_key => SecretKey,
+                                       region     => Region,
+                                       endpoint   => Endpoint}}),
+      setup_update_callback(Expiry, Ref, Role),
+      {reply, {ok, Ref}, State}
+  end;
+handle_call({add_creds, {static, AccessKey, SecretKey}, Opts}, _From, State) ->
+  case ets:lookup(creds_dedup, {static, AccessKey}) of
+    [{_, Ref}] -> {reply, {ok, Ref}, State};
+    [] ->
+      {Region, Endpoint} = parse_opts(Opts),
+      Ref = make_ref(),
+      true = ets:insert(creds_dedup, {{static, AccessKey}, Ref}),
+      true = ets:insert(creds, {Ref, #{access_key => AccessKey,
+                                       secret_key => SecretKey,
+                                       region     => Region,
+                                       endpoint   => Endpoint}}),
+      {reply, {ok, Ref}, State}
+  end;
+handle_call(Args, _From, State) ->
+  io:format("Unknown args ~p~n", [Args]),
```

Comment:
`error_logger:warning_msg("Unknown call ~p~n", [Args])` might be more appropriate and go to the built-in logger.